### ISSUE TITLE
[$1000] Migrate ReportActionCompose.js to function component

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#16263]
### Filenames
- [ ] [src/pages/home/report/ReportActionCompose.js](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose.js)
    - **type:** `React.Component`
    - **has state values:** true
    - **has refs:** true
    - **has context:** false
    - **lifecycle methods:** `componentDidMount`,`componentWillUnmount`,`componentDidUpdate`

### Task
- We currently have some class components in our codebase that we would like to refactor to a function component.
- Here's a link with some general advice on how to refactor a class component to a function component: https://react.dev/reference/react/Component#alternatives
- If you need additional guidance, please ask in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q)
- Test for any regressions and verify that there are no breaking changes

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01131d62c055beb52a</li>
        <li>Upwork Job ID: 1668304833402220544</li>
        <li>Last Price Increase: 2023-06-12</li>
<li>Automatic offers: </li>
<ul>
<li>0xmiroslav | Reviewer | 26454734</li>
<li>situchan | Contributor | 26454736</li>
</ul></li>
    </ul>
</details>

**Comment by roryabraham** [comment]
This one seems particularly nasty, so I'd like to take a stab. I'm pretty familiar with hooks but could use a refresher.

**Comment by marcaaron** [comment]
Heads up! The pricing for this issue has been adjusted based on the scope of the work and the fact that no complex bug investigations or proposal are required.

**Comment by hannojg** [comment]
Not sure if the issue is external but I can help with that refactor? (Need it to happen to unblock another PR 😄 )

**Comment by marcaaron** [comment]
We are going to open these up soon. Our internal team has been slowly chewing through the first round of 30 or so. But would love to have you work on this one or some others @hannojg !

**Comment by roryabraham** [comment]
Yeah, @hannojg I have a pretty big WIP [here](https://github.com/Expensify/App/pull/16978), and where I left off I was blocked because the lower-level `Composer` component also needs to be migrated in order for this to work.

**Comment by roryabraham** [comment]
I think @hannojg is going to take over my PR. @hannojg can you comment to confirm so I can assign this to you?

**Comment by hannojg** [comment]
Yes, here we go

**Comment by stas-kravchuk** [comment]
hey @roryabraham , the last few days I have been working on the ReportAcitionCompose.js migration, and the draft pr can be viewed [here](https://github.com/Expensify/App/pull/18648), the main migration has been finished and everything seems to be working fine at first sight. I wanted to ask, you mentioned above that you were blocked by the low-level `Composer` component which also needs to be migrated to get everything working, but I think I managed to migrate the component without `Composer` migrations
does it need to be migrated anyway, even if everything will work without migrating it?
P.S. PR still needs some polishing, but core migration has been finished, I've done some general testing on mobile and web and haven't noticed any obvious bugs or issues, but I'll keep looking

**Comment by chrispader** [comment]
@roryabraham i took over this issue from @stas-kravchuk. Are there any updates on his last question? Other than that, the component should be completely migrated and merged in `main` recently!

**Comment by roryabraham** [comment]
Answered in slack [here](https://expensify.slack.com/archives/C035J5C9FAP/p1684782399730379?thread_ts=1684404011.277179&cid=C035J5C9FAP) and [here](https://expensify.slack.com/archives/C035J5C9FAP/p1685562000090989?thread_ts=1685032152.962959&cid=C035J5C9FAP)

**Comment by marcaaron** [comment]
Taking this one off HOLD. But unsure what the status is.. @roryabraham can you let me know if this is still looking for help?

**Comment by muxriddinmuqimov77** [comment]
I'd like to work on this issue

**Comment by chrispader** [comment]
The [finished PR](https://github.com/Expensify/App/pull/18648) is already waiting for reviews! 😬🚀

**Comment by Vishrut19** [comment]
I want to work on it

**Comment by roryabraham** [comment]
Also, this is a more complicated migration and lots of testing is required, so I'm going to adjust the bounty to $1000

**Comment by roryabraham** [comment]
PR review is ongoing: https://github.com/Expensify/App/pull/18648

**Comment by dylanexpensify** [comment]
@0xmiroslav any update here my friend? 🙇

**Comment by chrispader** [comment]
We had some more conflicts, but i recently resolved them, so @0xmiroslav feel free to review! :)

https://github.com/Expensify/App/pull/18648#issuecomment-1631938579

**Comment by chrispader** [comment]
@0xmiroslav the PR is ready for review :)

**Comment by roryabraham** [comment]
@0xmiroslav I see that @hannojg has just cleaned up the PR and it's ready for another review

**Comment by dylanexpensify** [comment]
@0xmiroslav did we get this reviewed?

**Comment by dylanexpensify** [comment]
Didn't get an update today, will on Monday! Thanks @0xmiroslav! Welcome @situchan!

**Comment by dylanexpensify** [comment]
@situchan can we get an update

**Comment by situchan** [comment]
Feedback was addressed a few hrs ago and now PR is in final review

**Comment by dylanexpensify** [comment]
Waiting for PR to merge

**Comment by dylanexpensify** [comment]
Reviewing tomorrow! @0xmiroslav @situchan mind giving an update

**Comment by situchan** [comment]
PR deployed to staging

**Comment by dylanexpensify** [comment]
Nice - waiting for prod>no regressions

**Comment by dylanexpensify** [comment]
same ^

**Comment by dylanexpensify** [comment]
Should be today!

**Comment by dylanexpensify** [comment]
Hmm, Payment isn't autopopulating, but @marcaaron we're good to pay yeah?

**Comment by situchan** [comment]
> Hmm, Payment isn't autopopulating, but @marcaaron we're good to pay yeah?

@roryabraham is assigned engineer here.
PR was deployed to production on Aug 24 so yeah today is due.

It took much efforts to review 1.3k lines of code and do full regression test as this is the most important component in the app. Also there were tens of conflicts and had to retest whenever conflict fixed.
Hope that would be considered in payment.

**Comment by roryabraham** [comment]
Confirming that this was deployed to prod on Aug 24th so is ready to pay out.

Regarding payment, we actually had two C+ here. My proposal for payments is:

- @0xmiroslav, who reviewed several times but ended up having to pass the PR off. I think we should pay $500 to @0xmiroslav instead of $0 to account for the work they did in the earlier phases of the PR reviews.
- $1000 to @situchan (i.e: 4x the normal amount for functional component migrations)

I understand that this was a more time consuming PR to work on, and that's why I quadrupled the bounty early on here.

**Comment by roryabraham** [comment]
This is done – @dylanexpensify bump to pay plz

**Comment by dylanexpensify** [comment]
Paying tonight!
